odp_sdk/client.py

```diff
@@ -21,6 +22,105 @@
 log = logging.getLogger("odp-sdk.log")
 
 
+def not_none(value: Any) -> Any:
+    if value is None:
+        raise ValueError("Value should not be None")
+    return value
+
+
+class TokenHandler:
+    """Automatically obtain CogniteClient bearer tokens, checks expiry"""
+
+    TOKEN_EXPIRY_BUFFER_PERIOD_DEFAULT_SECONDS = 60.
+    DEFAULT_CDF_CLUSTER = "westeurope-1"
+    DEFAULT_LOGIN_URL = "https://oceandataplatform.b2clogin.com/oceandataplatform.onmicrosoft.com/B2C_1A_ROPC_Auth/oauth2/v2.0/token"
```

Comment:
I didn't think of this before, but the pipeline client is using Resource Owner Password Flow, which is a deamon/machine-machine type auth flow.

For users we should be using Authroization Code with PKCE, which involves redirecting the user to a B2C website where they login/create an account before they receive a token.

It's a 2 step process (from the code's perspective):
1. Call `https://oceandataplatform.b2clogin.com/oceandataplatform.onmicrosoft.com/b2c_1a_signup_signin_custom/oauth2/v2.0/authorize` with a callback url in params. This requires setting up a local server in order to handle the redirect from B2C that includes the exchange token.
2. Exchange the previous token with an access token from `https://oceandataplatform.b2clogin.com/oceandataplatform.onmicrosoft.com/b2c_1a_signup_signin_custom/oauth2/v2.0/token`

This means that username/password shouldn't be passed to the SDK for login, instead we should use Microsoft Authentication Library (MSAL).

My attempt at writing this in python is:

```py
from msal import PublicClientApplication

aad_configs = {
  'euwest1': {
        'authority': 'https://oceandataplatform.b2clogin.com/oceandataplatform.onmicrosoft.com/B2C_1A_signup_signin_custom',
        'client_id': 'b2a2d339-e785-4213-a773-8b289abd2199', // SDK AAD app-id
        'resource_uri': 'https://oceandataplatform.onmicrosoft.com/cdf-euwest1-api',
        'scopes': ["user_impersonation"],
    },
}

class OdpAuth:
    def __init__(self, project: str):
        self.__config = aad_configs[project]
        if self.__config == None:
            raise Exception('Invalid project')

        print('Initializing OdpAuth')
        print(self.__config)
        self.__msal_app = PublicClientApplication(
            client_id=self.__config['client_id'],
            authority=self.__config['authority']
        )
    
    def __authenticate(self, scopes):
        msal_app = self.__msal_app
        _scopes = self.__translateScopes(scopes)

        # Check if session already exists and re-use it if possible
        accounts = msal_app.get_accounts()
        if accounts:
            print("Authenticating silently")
            return msal_app.acquire_token_silent(scopes=_scopes, account=accounts[0])
    
        # interactive login - make sure you have http://localhost:port in Redirect URI in App Registration as type "Mobile and desktop applications"
        port = 53000
        creds = msal_app.acquire_token_interactive(scopes=_scopes, port=port)
        return creds

    def __prefix_scope(self, scope):
        recognized_scopes = ['DATA.VIEW','DATA.CHANGE','COMPUTE.VIEW','COMPUTE.CHANGE','ADMIN','user_impersonation','IDENTITY']
        # TODO: if (CDF config AND ...)
        if scope in recognized_scopes:
            return f"{self.__config['resource_uri']}/{scope}"
        return scope
    
    def __translateScopes(self, scopes):
        _scopes = scopes
        if isinstance(scopes, str):
            _scopes = " ".split(str)

        return [self.__prefix_scope(scope) for scope in _scopes]

    def get_access_token(self):
        return self.__authenticate(self.__config['scopes'])
```